### ISSUE TITLE
Clarify stages of MIR pipeline

### ISSUE THREAD

**Issue by ecstatic-morse** [rust-lang/rust#72515]
The [`rustc_mir::transform`](https://github.com/rust-lang/rust/blob/master/src/librustc_mir/transform/mod.rs) module defines the pipeline of analysis passes and transformations that  run on the MIR after is lowered but before it is passed to codegen. However, it's not always clear whether certain invariants apply to each stage of the pipeline. For example, some terminators (`FalseEdges`, `DropAndReplace`) are removed entirely at each stage of the pipeline and the `Return` terminator, which initially appears only once in the entire MIR body may appear multiple times after the `generator` transform.

Additionally, it can be hard to find the right place for new error-emitting passes like the lint in #72270 or #71824. Is there a point at which the MIR is no longer meant for analysis but only for codegen? Finally, the names of the `mir_*` queries have lost their meaning over time: `mir_const` can probably be removed entirely, and `optimized_mir` and `promoted_mir` should probably be renamed to `mir_optimized` and `mir_promoted_fragments_optimized` respectively.

**Comment by RalfJung** [comment]
> the Return terminator, which initially appears only once in the entire MIR body may appear multiple times after the generator transform.

Ah I didn't know this; this should also be reflected in the docs for `TerminatorKind::Return`.

**Comment by RalfJung** [comment]
> `mir_promoted_fragments_optimized`

I don't think I understand what this name means.  What would be the doc comment that goes with it?

**Comment by ecstatic-morse** [comment]
`promoted_mir` doesn't do promotion (I believe that happens in `mir_validated`?). It steals the promoted MIR and optimizes it.

**Comment by RalfJung** [comment]
Why does it need to be a separate query?

**Comment by RalfJung** [comment]
@oli-obk added infrastructure to MIR validation to distinguish different "stages".

But we still do not have a clear place to put MIR lints, and we currently do this inconsistently. https://github.com/rust-lang/rust/pull/72270 added a "MIR transform" that is just a lint; there is [a `lints::check` call](https://github.com/rust-lang/rust/blob/31add7e60709445617ab54a69f6f21cfcb2e3122/src/librustc_mir_build/build/mod.rs#L184) in MIR construction, and unsafety checking is its own whole thing.

**Comment by RalfJung** [comment]
Also Cc @rust-lang/wg-mir-opt for the stakeholders and code owner questions in the OP.

**Comment by camsteffen** [comment]
Is this solved by #91475?

**Comment by RalfJung** [comment]
I don't think so. We still have mir_const, mir_promoted, mir_for_ctfe, mir_drops_elaborated_and_const_checked, run_optimization_passes.

For lints, the situation seems to have improved -- we now have MirLint, that can be turned into passes. On the other hand, there also still is some mention of lints during MIR building, and check_unsafety is its own special thing and not a MIR pass.

**Comment by RalfJung** [comment]
Cc @compiler-errors -- not sure how you feel about the status quo here, I just saw you do a bit of cleanup in https://github.com/rust-lang/rust/pull/135705. I feel like the the "mir_const_qualif, mir_promoted, mir_for_ctfe, mir_drops_elaborated_and_const_checked" chain is still somewhat messy to understand.
